store/models/service_agreement.go

```diff
@@ -33,25 +35,53 @@ type Signer interface {
 }
 
 // NewServiceAgreementFromRequest builds a new ServiceAgreement.
-func NewServiceAgreementFromRequest(sar ServiceAgreementRequest, signer Signer) (ServiceAgreement, error) {
-	id, err := generateServiceAgreementID(sar.Encumbrance, sar.Digest)
+func NewServiceAgreementFromRequest(reader io.Reader, signer Signer) (ServiceAgreement, error) {
+	var sa ServiceAgreement
+	input, err := ioutil.ReadAll(reader)
 	if err != nil {
-		return ServiceAgreement{}, err
+		return sa, err
+	}
+
+	var sar ServiceAgreementRequest
+	err = json.Unmarshal(input, &sar)
+	if err != nil {
+		return sa, err
+	}
+
+	normalized, err := utils.NormalizedJSON(input)
+	if err != nil {
+		return sa, err
+	}
+
+	requestDigest, err := utils.Keccak256([]byte(normalized))
+	digest := common.ToHex(requestDigest)
+
+	encumbrance := Encumbrance{Payment: sar.Payment, Expiration: sar.Expiration}
+
+	id, err := generateServiceAgreementID(encumbrance, digest)
+	if err != nil {
+		return sa, err
 	}
 
 	signature, err := signer.Sign([]byte(id))
 	if err != nil {
-		return ServiceAgreement{}, err
+		return sa, err
 	}
 
-	return ServiceAgreement{
-		CreatedAt:   Time{time.Now()},
-		Encumbrance: sar.Encumbrance,
-		ID:          id,
-		jobSpec:     sar.JobSpec,
-		RequestBody: sar.NormalizedBody,
-		Signature:   signature,
-	}, err
+	sa.ID = id
+	sa.CreatedAt = Time{time.Now()}
+	sa.Encumbrance = encumbrance
+
+	sa.jobSpec = NewJob()
+	sa.jobSpec.Initiators = sar.Initiators
```

Comment:
I understand that you're beholden to the `NewJob()` constructor and can't do the hash style initialization, but we could still do it for `sa` which would be more idiomatic go:

```golang
sa = ServiceAgreement{
  ID: id,
  jobSpec: previouslyInitializedJobSpec,
  RequestBody: normalized,
  ...
}
return sa, err
```

Reply:
Right, which I believe more explicitly conveys you're returning a 0 value inline rather than having to scroll up and find the reference value.

Reply:
Hrmm... I like it. Will change.